python/dgl/backend/backend.py

```diff
@@ -842,6 +833,105 @@ def zerocopy_from_numpy(np_array):
     """
     pass
 
+def zerocopy_to_dgl_ndarray(input):
+    """Zerocopy a framework-specific Tensor to dgl.ndarray.NDArray
+
+    Parameters
+    ----------
+    input : Tensor
+
+    Returns
+    -------
+    dgl.ndarray.NDArray
+    """
+    pass
+
+def zerocopy_from_dgl_ndarray(input):
+    """Zerocopy a dgl.ndarray.NDArray to framework-specific Tensor
+
+    Parameters
+    ----------
+    input : dgl.ndarray.NDArray
+
+    Returns
+    -------
+    Tensor
+    """
+    pass
+
+###############################################################################
+# Custom Operators for graph level computations.
+
+# Note: These operators are supposed to be implemented using DGL-provided
+# kernels (see kernel.py), and plug into tensor framework using custom op
+# extensions.
+
+def binary_reduce(reducer, binary_op, graph, lhs, rhs, lhs_data, rhs_data,
+                  out_size, lhs_map, rhs_map, out_map):
+    """Perform binary operation between given data and reduce based on graph
+    structure.
+
+    Parameters
+    ----------
+    reducer : str
+        Type of reduction: 'sum', 'max', 'min', 'mean', 'prod', 'none' (no
+        reduction)
+    binary_op : str
+        Binary operation to perform, can be 'add', 'mul', 'sub', 'div'
+    graph : GraphIndex
+        The graph
+    lhs : int
+        The lhs target (src, dst, edge)
+    rhs : int
+        The rhs target (src, dst, edge)
+    lhs_data : Tensor
+        The lhs data
+    rhs_data : Tensor
+        The rhs data
+    out_size : int
+        Size of first dimension of output data
+    lhs_map : tuple of two dgl.ndarray.NDArray
+        The lhs id mapping array
+    rhs_map : tuple of two dgl.ndarray.NDArray
+        The rhs id mapping array
+    out_map : tuple of two dgl.ndarray.NDArray
+        The out id mapping array
```

Comment:
Why are they tuples?

Reply:
Because the mapping for forward pass and backward pass can possibly be different. It's the same for nodes, but different for edges because when converting coo to csr and csc, the edge order is different.